pkg/controller/servicebindingrequest/reconciler.go

```diff
@@ -33,57 +34,89 @@ const (
 	requeueAfter int64 = 45
 )
 
-// setSecretName update the CR status field to "in progress", and setting secret name.
-func (r *Reconciler) setSecretName(
-	ctx context.Context,
-	instance *v1alpha1.ServiceBindingRequest,
-	name string,
-) error {
-	instance.Status.BindingStatus = bindingInProgress
-	instance.Status.Secret = name
-	return r.client.Status().Update(ctx, instance)
+// getServiceBindingRequest retrieve the SBR object based on namespaced-name.
+func (r *Reconciler) getServiceBindingRequest(
+	namespacedName types.NamespacedName,
+) (*v1alpha1.ServiceBindingRequest, error) {
+	gr := v1alpha1.SchemeGroupVersion.WithResource(ServiceBindingRequestResource)
+	resourceClient := r.dynClient.Resource(gr).Namespace(namespacedName.Namespace)
+	u, err := resourceClient.Get(namespacedName.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	sbr := &v1alpha1.ServiceBindingRequest{}
+	err = runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, sbr)
+	if err != nil {
+		return nil, err
+	}
+	return sbr, nil
+}
+
+// setStatusServiceBindingRequest update the status of a SBR object.
+func (r *Reconciler) setStatusServiceBindingRequest(sbr *v1alpha1.ServiceBindingRequest) error {
+	data, err := runtime.DefaultUnstructuredConverter.ToUnstructured(sbr)
+	if err != nil {
+		return err
+	}
+	u := &unstructured.Unstructured{Object: data}
 
+	gr := v1alpha1.SchemeGroupVersion.WithResource(ServiceBindingRequestResource)
+	resourceClient := r.dynClient.Resource(gr).Namespace(sbr.GetNamespace())
+	_, err = resourceClient.UpdateStatus(u, metav1.UpdateOptions{})
+	return err
 }
 
-// setStatus update the CR status field.
-func (r *Reconciler) setStatus(
-	ctx context.Context,
-	instance *v1alpha1.ServiceBindingRequest,
-	status string,
-) error {
-	instance.Status.BindingStatus = status
-	return r.client.Status().Update(ctx, instance)
+// setSecretName update the CR status field to "in progress", and setting secret name.
+func (r *Reconciler) setSecretName(namespacedName types.NamespacedName, name string) error {
+	sbr, err := r.getServiceBindingRequest(namespacedName)
```

Comment:
I have a concern here. If the SBR has changed after the reconcile was triggered, in that case, it is fair to have the `update` fail, and let the next re-queue take care if it. This is specific to the SBR since our controller manages that and we know that nothing on the controller side would be messing with the object.

Reply:
Looks like we should have the latest object on updating the status because the returned value is copied over into the `instance` reference by design, right?
https://github.com/redhat-developer/service-binding-operator/blob/master//pkg/controller/servicebindingrequest/reconciler.go#L49:22

What do you think?

Reply:
@sbose78, I think we should be updating the `status`in SBR only once, and only at the end of the workflow, that would give a more clear approach in terms of code, and avoid this type of implementation all together. :-)

Reply:
> Looks like we should have the latest object on updating the status because the returned value is copied over into the `instance` reference by design, right?
> https://github.com/redhat-developer/service-binding-operator/blob/master//pkg/controller/servicebindingrequest/reconciler.go#L49:22

I don't see a difference in approach here. I moved the part that fetches the actual instance of an SBR to a method, and reused that method in different places.
Before, the value was copied over to `instance` using the same approach, since the `Reconciler` will only received a namespaced-named of an SBR, nothing more than that.

Reply:
>  I think we should be updating the statusin SBR only once, and only at the end of the workflow

+1 I'm good with that!